File: deepeval/metrics/dag/serialization/serialization.py
"""JSON serialization for ``DeepAcyclicGraph``.

The JSON document describes ONLY graph structure. It does NOT encode mode
(single-turn vs multiturn), version, or root list - those are inferred or
supplied by the caller.

JSON shape::

    {
      "nodes": {
        "<id>": {
          "type": "TaskNode" | "BinaryJudgementNode" | ...,
          ... node-specific fields,
          "children": ["<id>", ...]                # for non-VerdictNode
        },
        "<id>": {
          "type": "VerdictNode",
          "verdict": <bool|str>,
          "score": <int>            # XOR with "child"
          | "child": {              # see ChildType for the discriminator
              "type": "node",   "ref": "<id>"
            | "type": "geval",  ...constructor kwargs
            | "type": "metric", "metric_class": "<class name>", "kwargs": {...}
          }
        }
      }
    }
"""

from __future__ import annotations

import importlib
import inspect
import json
import uuid
from collections import deque
from enum import Enum
from typing import Any, Dict, List, Optional, Set, Type

from deepeval.metrics.base_metric import BaseConversationalMetric, BaseMetric
from deepeval.metrics.conversational_dag.nodes import (
    ConversationalBaseNode,
    ConversationalBinaryJudgementNode,
    ConversationalNonBinaryJudgementNode,
    ConversationalTaskNode,
    ConversationalVerdictNode,
)
from deepeval.metrics.conversational_g_eval.conversational_g_eval import (
    ConversationalGEval,
)
from deepeval.metrics.dag.nodes import (
    BaseNode,
    BinaryJudgementNode,
    NonBinaryJudgementNode,
    TaskNode,
    VerdictNode,
)
from deepeval.metrics.g_eval.g_eval import GEval
from deepeval.test_case import LLMTestCaseParams, TurnParams

from .registry import CLASS_TO_NODE_TYPE, NODE_CLASSES
from .types import ChildType, NodeType


# ----------------------------------------------------------------------------
# Public API
# ----------------------------------------------------------------------------


def dag_to_dict(dag) -> Dict[str, Any]:
    """Serialize a ``DeepAcyclicGraph`` instance to a JSON-friendly dict."""
    from deepeval.metrics.dag.utils import is_valid_dag_from_roots

    if not is_valid_dag_from_roots(
        root_nodes=dag.root_nodes, multiturn=dag.multiturn
    ):
        raise ValueError("Cycle detected in DAG graph; cannot serialize.")

    ordered_nodes = _walk_nodes(dag.root_nodes)
    id_map = _assign_ids(ordered_nodes)

    nodes_dict: Dict[str, Dict[str, Any]] = {}
    for node in ordered_nodes:
        node_id = id_map[id(node)]
        nodes_dict[node_id] = _serialize_node(node, id_map)

    return {"nodes": nodes_dict}


def dag_to_json(dag, indent: int = 2) -> str:
    return json.dumps(dag_to_dict(dag), indent=indent)


def dag_from_dict(data: Dict[str, Any], multiturn: bool = False):
    """Re-create a ``DeepAcyclicGraph`` from a dict produced by ``dag_to_dict``."""
    from deepeval.metrics.dag.graph import DeepAcyclicGraph

    if not isinstance(data, dict) or "nodes" not in data:
        raise ValueError(
            "Invalid DAG document: expected a top-level object with a 'nodes' key."
        )
    nodes_spec = data["nodes"]
    if not isinstance(nodes_spec, dict) or len(nodes_spec) == 0:
        raise ValueError(
            "Invalid DAG document: 'nodes' must be a non-empty object."
        )

    # Coerce/validate every node 'type' up front for clear errors.
    for node_id, spec in nodes_spec.items():
        if not isinstance(spec, dict) or "type" not in spec:
            raise ValueError(
                f"Node '{node_id}' is missing required 'type' field."
            )
        try:
            NodeType(spec["type"])
        except ValueError:
            valid = ", ".join(nt.value for nt in NodeType)
            raise ValueError(
                f"Node '{node_id}' has unknown type '{spec['type']}'. "
                f"Expected one of: {valid}."
            )

    referenced = _collect_referenced_ids(nodes_spec)
    root_ids = [nid for nid in nodes_spec.keys() if nid not in referenced]
    if not root_ids:
        raise ValueError(
            "No root nodes detected (every node is referenced as a child); "
            "graph would be empty or contain a cycle."
        )

    class_map = NODE_CLASSES[bool(multiturn)]
    built: Dict[str, Any] = {}

    def build(node_id: str, stack: Set[str]):
        if node_id in built:
            return built[node_id]
        if node_id in stack:
            raise ValueError(
                f"Cycle detected in JSON refs involving node '{node_id}'."
            )
        if node_id not in nodes_spec:
            raise ValueError(f"Reference to unknown node id '{node_id}'.")

        stack.add(node_id)
        spec = nodes_spec[node_id]
        nt = NodeType(spec["type"])
        cls = class_map[nt]
        node: Any

        if nt == NodeType.VERDICT:
            node = _build_verdict(spec, cls, multiturn, build, stack)
        elif nt == NodeType.TASK:
            children = [build(cid, stack) for cid in spec.get("children", [])]
            kwargs = _task_kwargs(spec, multiturn)
            node = cls(children=children, **kwargs)
        elif nt in (NodeType.BINARY_JUDGEMENT, NodeType.NON_BINARY_JUDGEMENT):
            children = [build(cid, stack) for cid in spec.get("children", [])]
            kwargs = _judgement_kwargs(spec, multiturn)
            node = cls(children=children, **kwargs)
        else:
            raise ValueError(f"Unhandled node type '{nt}'.")

        stack.discard(node_id)
        built[node_id] = node
        return node

    root_nodes = [build(rid, set()) for rid in root_ids]
    return DeepAcyclicGraph(root_nodes=root_nodes)


def dag_from_json(s: str, multiturn: bool = False):
    return dag_from_dict(json.loads(s), multiturn=multiturn)


# ----------------------------------------------------------------------------
# Serialization helpers
# ----------------------------------------------------------------------------


def _walk_nodes(root_nodes: List[Any]) -> List[Any]:
    """BFS from each root, returning every reachable node exactly once,
    in stable BFS order (roots first)."""
    seen: Set[int] = set()
    ordered: List[Any] = []
    queue: deque = deque(root_nodes)
    while queue:
        node = queue.popleft()
        if id(node) in seen:
            continue
        seen.add(id(node))
        ordered.append(node)

        for child in _iter_children(node):
            queue.append(child)
    return ordered


def _iter_children(node: Any):
    if hasattr(node, "children") and node.children:
        for c in node.children:
            yield c
    if isinstance(node, (VerdictNode, ConversationalVerdictNode)):
        if node.child is not None and _is_node(node.child):
            yield node.child


def _is_node(obj: Any) -> bool:
    return isinstance(obj, (BaseNode, ConversationalBaseNode))


def _assign_ids(ordered_nodes: List[Any]) -> Dict[int, str]:
    """Assign a fresh uuid4 string to every node, keyed by id(node)."""
    return {id(node): str(uuid.uuid4()) for node in ordered_nodes}


def _serialize_node(node: Any, id_map: Dict[int, str]) -> Dict[str, Any]:
    cls = type(node)
    if cls not in CLASS_TO_NODE_TYPE:
        raise ValueError(
            f"Unsupported node class '{cls.__name__}'; cannot serialize."
        )
    nt = CLASS_TO_NODE_TYPE[cls]

    if nt == NodeType.TASK:
        out: Dict[str, Any] = {
            "type": nt.value,
            "instructions": node.instructions,
            "output_label": node.output_label,
            "label": node.label,
            "evaluation_params": _serialize_eval_params(node.evaluation_params),
            "children": [id_map[id(c)] for c in node.children],
        }
        if (
            isinstance(node, ConversationalTaskNode)
            and node.turn_window is not None
        ):
            out["turn_window"] = list(node.turn_window)
        return out

    if nt in (NodeType.BINARY_JUDGEMENT, NodeType.NON_BINARY_JUDGEMENT):
        out = {
            "type": nt.value,
            "criteria": node.criteria,
            "label": node.label,
            "evaluation_params": _serialize_eval_params(node.evaluation_params),
            "children": [id_map[id(c)] for c in node.children],
        }
        if (
            isinstance(
                node,
                (
                    ConversationalBinaryJudgementNode,
                    ConversationalNonBinaryJudgementNode,
                ),
            )
            and node.turn_window is not None
        ):
            out["turn_window"] = list(node.turn_window)
        return out

    if nt == NodeType.VERDICT:
        out = {"type": nt.value, "verdict": node.verdict}
        if node.score is not None:
            out["score"] = node.score
        if node.child is not None:
            out["child"] = _serialize_verdict_child(node.child, id_map)
        return out

    raise ValueError(f"Unhandled node type '{nt}'.")  # pragma: no cover


def _serialize_eval_params(params) -> Optional[List[str]]:
    if params is None:
        return None
    return [p.value for p in params]


def _serialize_verdict_child(
    child: Any, id_map: Dict[int, str]
) -> Dict[str, Any]:
    if _is_node(child):
        return {"type": ChildType.NODE.value, "ref": id_map[id(child)]}
    if isinstance(child, (GEval, ConversationalGEval)):
        return _serialize_geval(child)
    if isinstance(child, (BaseMetric, BaseConversationalMetric)):
        return _serialize_metric(child)
    raise ValueError(
        f"VerdictNode.child has unsupported type '{type(child).__name__}'. "
        "Expected a BaseNode, GEval/ConversationalGEval, or a BaseMetric/"
        "BaseConversationalMetric subclass."
    )


def _serialize_geval(geval: Any) -> Dict[str, Any]:
    init_params = _init_param_names(type(geval))
    payload: Dict[str, Any] = {"type": ChildType.GEVAL.value}
    for name in init_params:
        if name == "self":
            continue
        if not hasattr(geval, name):
            continue
        value = getattr(geval, name)
        if name == "evaluation_params":
            payload[name] = _serialize_eval_params(value)
            continue
        json_value = _maybe_jsonify(value)
        if json_value is _SKIP:
            continue
        payload[name] = json_value
    return payload


def _serialize_metric(metric: Any) -> Dict[str, Any]:
    cls = type(metric)
    init_params = _init_param_names(cls)
    kwargs: Dict[str, Any] = {}
    for name in init_params:
        if name == "self":
            continue
        if not hasattr(metric, name):
            continue
        value = getattr(metric, name)
        if name == "evaluation_params":
            serialized = _serialize_eval_params(value)
            if serialized is not None:
                kwargs[name] = serialized
            continue
        json_value = _maybe_jsonify(value)
        if json_value is _SKIP:
            continue
        kwargs[name] = json_value

    return {
        "type": ChildType.METRIC.value,
        "metric_class": cls.__name__,
        "kwargs": kwargs,
    }


def _init_param_names(cls: Type) -> List[str]:
    try:
        sig = inspect.signature(cls.__init__)
    except (TypeError, ValueError):
        return []
    return list(sig.parameters.keys())


_SKIP = object()


def _maybe_jsonify(value: Any) -> Any:
    """Return a JSON-friendly version of ``value`` or ``_SKIP`` if it cannot
    be safely round-tripped."""
    if value is None:
        return None
    if isinstance(value, (bool, int, float, str)):
        return value
    if isinstance(value, (list, tuple)):
        out: List[Any] = []
        for item in value:
            jv = _maybe_jsonify(item)
            if jv is _SKIP:
                return _SKIP
            out.append(jv)
        return out
    if isinstance(value, dict):
        out_d: Dict[str, Any] = {}
        for k, v in value.items():
            if not isinstance(k, str):
                return _SKIP
            jv = _maybe_jsonify(v)
            if jv is _SKIP:
                return _SKIP
            out_d[k] = jv
        return out_d
    if isinstance(value, Enum):
        return _maybe_jsonify(value.value)
    # Anything else (DeepEvalBaseLLM instances, classes, callables, custom
    # objects, etc.) is skipped.
    return _SKIP


# ----------------------------------------------------------------------------
# Deserialization helpers
# ----------------------------------------------------------------------------


def _collect_referenced_ids(nodes_spec: Dict[str, Any]) -> Set[str]:
    referenced: Set[str] = set()
    for spec in nodes_spec.values():
        for cid in spec.get("children", []) or []:
            referenced.add(cid)
        child = spec.get("child")
        if (
            isinstance(child, dict)
            and child.get("type") == ChildType.NODE.value
        ):
            ref = child.get("ref")
            if isinstance(ref, str):
                referenced.add(ref)
    return referenced


def _eval_params_cls(multiturn: bool):
    return TurnParams if multiturn else LLMTestCaseParams


def _deserialize_eval_params(values, multiturn: bool):
    if values is None:
        return None
    enum_cls = _eval_params_cls(multiturn)
    out = []
    for v in values:
        try:
            out.append(enum_cls(v))
        except ValueError:
            valid = ", ".join(p.value for p in enum_cls)
            raise ValueError(
                f"Unknown evaluation_param '{v}'. Expected one of: {valid}."
            )
    return out


def _task_kwargs(spec: Dict[str, Any], multiturn: bool) -> Dict[str, Any]:
    kwargs: Dict[str, Any] = {
        "instructions": spec["instructions"],
        "output_label": spec["output_label"],
    }
    if "label" in spec and spec["label"] is not None:
        kwargs["label"] = spec["label"]
    if "evaluation_params" in spec:
        kwargs["evaluation_params"] = _deserialize_eval_params(
            spec["evaluation_params"], multiturn
        )
    if multiturn and "turn_window" in spec and spec["turn_window"] is not None:
        kwargs["turn_window"] = tuple(spec["turn_window"])
    return kwargs


def _judgement_kwargs(spec: Dict[str, Any], multiturn: bool) -> Dict[str, Any]:
    kwargs: Dict[str, Any] = {"criteria": spec["criteria"]}
    if "label" in spec and spec["label"] is not None:
        kwargs["label"] = spec["label"]
    if "evaluation_params" in spec:
        kwargs["evaluation_params"] = _deserialize_eval_params(
            spec["evaluation_params"], multiturn
        )
    if multiturn and "turn_window" in spec and spec["turn_window"] is not None:
        kwargs["turn_window"] = tuple(spec["turn_window"])
    return kwargs


def _build_verdict(
    spec: Dict[str, Any],
    cls: Type,
    multiturn: bool,
    build,
    stack: Set[str],
):
    verdict = spec["verdict"]
    if "score" in spec and spec["score"] is not None:
        return cls(verdict=verdict, score=spec["score"])
    if "child" not in spec or spec["child"] is None:
        raise ValueError(
            "VerdictNode spec must have either 'score' or 'child'."
        )
    child_spec = spec["child"]
    if not isinstance(child_spec, dict) or "type" not in child_spec:
        raise ValueError(
            "VerdictNode 'child' must be an object with a 'type' field."
        )
    try:
        ctype = ChildType(child_spec["type"])
    except ValueError:
        valid = ", ".join(c.value for c in ChildType)
        raise ValueError(
            f"VerdictNode child has unknown type '{child_spec['type']}'. "
            f"Expected one of: {valid}."
        )

    if ctype == ChildType.NODE:
        ref = child_spec.get("ref")
        if not isinstance(ref, str):
            raise ValueError("VerdictNode child of type 'node' requires 'ref'.")
        child_obj = build(ref, stack)
    elif ctype == ChildType.GEVAL:
        child_obj = _build_geval(child_spec, multiturn)
    else:
        child_obj = _build_metric(child_spec)

    return cls(verdict=verdict, child=child_obj)


def _build_geval(child_spec: Dict[str, Any], multiturn: bool):
    cls = ConversationalGEval if multiturn else GEval
    kwargs = {k: v for k, v in child_spec.items() if k != "type"}
    if "evaluation_params" in kwargs:
        kwargs["evaluation_params"] = _deserialize_eval_params(
            kwargs["evaluation_params"], multiturn
        )
    return cls(**kwargs)


def _build_metric(child_spec: Dict[str, Any]):
    metric_class = child_spec.get("metric_class")
    if not isinstance(metric_class, str) or not metric_class:
        raise ValueError(
            "Metric child requires a non-empty 'metric_class' field."
        )
    metrics_module = importlib.import_module("deepeval.metrics")
    cls = getattr(metrics_module, metric_class, None)
    if cls is None:
        raise ValueError(
            f"Unknown metric_class '{metric_class}'. "
            f"It must be importable from 'deepeval.metrics'."
        )
    kwargs = dict(child_spec.get("kwargs") or {})
    # Reconstruct evaluation_params enum list if present.
    if "evaluation_params" in kwargs and isinstance(
        kwargs["evaluation_params"], list
    ):
        # Try LLMTestCaseParams first, then TurnParams for conversational metrics.
        if issubclass(cls, BaseConversationalMetric):
            kwargs["evaluation_params"] = _deserialize_eval_params(
                kwargs["evaluation_params"], multiturn=True
            )
        else:
            kwargs["evaluation_params"] = _deserialize_eval_params(
                kwargs["evaluation_params"], multiturn=False
            )
    return cls(**kwargs)

File: typescript/src/metrics/dag/serialization.ts
import { randomUUID } from "node:crypto";
import { MultiTurnParams, SingleTurnParams } from "@/test-case";
import { GEval } from "@/metrics/g-eval/g-eval";
import { ConversationalGEval } from "@/metrics/conversational-g-eval/conversational-g-eval";
import {
  BinaryJudgementNode,
  NonBinaryJudgementNode,
  TaskNode,
  VerdictNode,
} from "@/metrics/dag/nodes";
import {
  ConversationalBinaryJudgementNode,
  ConversationalNonBinaryJudgementNode,
  ConversationalTaskNode,
  ConversationalVerdictNode,
  type TurnWindow,
} from "@/metrics/conversational-dag/nodes";
import { DeepAcyclicGraph } from "@/metrics/dag/graph";
import {
  type AnyNode,
  type ChildMetric,
  type ExecutableNode,
  type VerdictLikeNode,
  isNode,
  isVerdictNode,
} from "@/metrics/dag/types";
import { edgesOf, isValidDagFromRoots } from "@/metrics/dag/utils";

export enum NodeType {
  TASK = "TaskNode",
  BINARY_JUDGEMENT = "BinaryJudgementNode",
  NON_BINARY_JUDGEMENT = "NonBinaryJudgementNode",
  VERDICT = "VerdictNode",
}

export enum ChildType {
  NODE = "node",
  GEVAL = "geval",
  METRIC = "metric",
}

const NODE_TYPES: Record<AnyNode["nodeKind"], NodeType> = {
  task: NodeType.TASK,
  binaryJudgement: NodeType.BINARY_JUDGEMENT,
  nonBinaryJudgement: NodeType.NON_BINARY_JUDGEMENT,
  verdict: NodeType.VERDICT,
};

type NodeConstructors = Record<NodeType, any>;

const NODE_CLASSES: Record<"single" | "multi", NodeConstructors> = {
  single: {
    [NodeType.TASK]: TaskNode,
    [NodeType.BINARY_JUDGEMENT]: BinaryJudgementNode,
    [NodeType.NON_BINARY_JUDGEMENT]: NonBinaryJudgementNode,
    [NodeType.VERDICT]: VerdictNode,
  },
  multi: {
    [NodeType.TASK]: ConversationalTaskNode,
    [NodeType.BINARY_JUDGEMENT]: ConversationalBinaryJudgementNode,
    [NodeType.NON_BINARY_JUDGEMENT]: ConversationalNonBinaryJudgementNode,
    [NodeType.VERDICT]: ConversationalVerdictNode,
  },
};

/**
 * Metric classes a `metric`-typed verdict child can be rebuilt from. TypeScript
 * has no import-by-name, so anything beyond G-Eval must be registered by the
 * caller before deserializing.
 */
export const METRIC_CLASS_REGISTRY: Record<string, new (options: any) => any> =
  {
    GEval,
    ConversationalGEval,
  };

export function registerMetricClass(
  name: string,
  metricClass: new (options: any) => any,
): void {
  METRIC_CLASS_REGISTRY[name] = metricClass;
}

// ---------------------------------------------------------------- serializing

/** BFS from each root, every reachable node exactly once, roots first. */
export function walkNodes(rootNodes: AnyNode[]): AnyNode[] {
  const seen = new Set<AnyNode>();
  const ordered: AnyNode[] = [];
  const queue = [...rootNodes];
  while (queue.length > 0) {
    const node = queue.shift()!;
    if (seen.has(node)) continue;
    seen.add(node);
    ordered.push(node);
    queue.push(...edgesOf(node));
  }
  return ordered;
}

function assignIds(nodes: AnyNode[]): Map<AnyNode, string> {
  return new Map(nodes.map((node) => [node, randomUUID()]));
}

/**
 * Params travel snake_cased so a DAG serialized here can be read by the Python
 * SDK, whose enum values are the snake_case spellings of these same fields.
 */
function toWireParam(value: string): string {
  return value.replace(/[A-Z]/g, (char) => `_${char.toLowerCase()}`);
}

function fromWireParam(value: string): string {
  return value.replace(/_([a-z])/g, (_, char: string) => char.toUpperCase());
}

function serializeEvalParams(
  params?: Array<SingleTurnParams | MultiTurnParams>,
): string[] | null {
  return params == null ? null : params.map((param) => toWireParam(param));
}

/** Public metric options we can round-trip; anything else is dropped. */
function serializeMetricOptions(metric: any): Record<string, unknown> {
  const options: Record<string, unknown> = {};
  const put = (key: string, value: unknown) => {
    if (value != null) options[key] = value;
  };
  put("name", metric.metricName);
  put("criteria", metric.criteria);
  put("evaluation_steps", metric.evaluationSteps);
  put("evaluation_params", serializeEvalParams(metric.evaluationParams));
  put("threshold", metric.threshold);
  put("strict_mode", metric.strictMode);
  return options;
}

function serializeVerdictChild(
  child: AnyNode | ChildMetric,
  ids: Map<AnyNode, string>,
): Record<string, unknown> {
  if (isNode(child)) {
    return { type: ChildType.NODE, ref: ids.get(child) };
  }
  if (child instanceof GEval || child instanceof ConversationalGEval) {
    return { type: ChildType.GEVAL, ...serializeMetricOptions(child) };
  }
  return {
    type: ChildType.METRIC,
    metric_class: child.constructor.name,
    kwargs: serializeMetricOptions(child),
  };
}

type ChildSerializer = (
  child: AnyNode | ChildMetric,
  ids: Map<AnyNode, string>,
) => Record<string, unknown>;

function serializeNode(
  node: AnyNode,
  ids: Map<AnyNode, string>,
  serializeChild: ChildSerializer,
): Record<string, unknown> {
  const type = NODE_TYPES[node.nodeKind];

  if (isVerdictNode(node)) {
    const out: Record<string, unknown> = { type, verdict: node.verdict };
    if (node.score != null) out.score = node.score;
    if (node.child != null) {
      out.child = serializeChild(node.child, ids);
    }
    return out;
  }

  const turnWindow = (node as { turnWindow?: TurnWindow }).turnWindow;
  const out: Record<string, unknown> = {
    type,
    label: node.label ?? null,
    evaluation_params: serializeEvalParams(node.evaluationParams),
    children: node.children.map((child) => ids.get(child)),
  };
  if (node.nodeKind === "task") {
    const task = node as unknown as TaskNode;
    out.instructions = task.instructions;
    out.output_label = task.outputLabel;
  } else {
    out.criteria = (node as unknown as BinaryJudgementNode).criteria;
  }
  if (turnWindow != null) out.turn_window = [...turnWindow];
  return out;
}

function serializeDag(
  dag: DeepAcyclicGraph,
  serializeChild: ChildSerializer,
): Record<string, unknown> {
  if (!isValidDagFromRoots(dag.rootNodes)) {
    throw new Error("Cycle detected in DAG graph; cannot serialize.");
  }
  const ordered = walkNodes(dag.rootNodes);
  const ids = assignIds(ordered);
  const nodes: Record<string, unknown> = {};
  for (const node of ordered) {
    nodes[ids.get(node)!] = serializeNode(node, ids, serializeChild);
  }
  return { nodes };
}

export function dagToDict(dag: DeepAcyclicGraph): Record<string, unknown> {
  return serializeDag(dag, serializeVerdictChild);
}

export function dagToJson(dag: DeepAcyclicGraph, indent = 2): string {
  return JSON.stringify(dagToDict(dag), null, indent);
}

// -------------------------------------------------------------- deserializing

function collectReferencedIds(nodesSpec: Record<string, any>): Set<string> {
  const referenced = new Set<string>();
  for (const spec of Object.values(nodesSpec)) {
    for (const id of spec.children ?? []) referenced.add(id);
    if (
      spec.child?.type === ChildType.NODE &&
      typeof spec.child.ref === "string"
    ) {
      referenced.add(spec.child.ref);
    }
  }
  return referenced;
}

function deserializeEvalParams(
  values: unknown,
  multiturn: boolean,
): Array<SingleTurnParams | MultiTurnParams> | undefined {
  if (values == null) return undefined;
  const valid: string[] = Object.values(
    multiturn ? MultiTurnParams : SingleTurnParams,
  );
  return (values as string[]).map((value) => {
    const param = fromWireParam(value);
    if (!valid.includes(param)) {
      throw new Error(
        `Unknown evaluation param '${value}'. Expected one of: ${valid
          .map(toWireParam)
          .join(", ")}.`,
      );
    }
    return param as SingleTurnParams | MultiTurnParams;
  });
}

function nodeOptions(spec: any, multiturn: boolean): Record<string, unknown> {
  const options: Record<string, unknown> = {
    label: spec.label ?? undefined,
    evaluationParams: deserializeEvalParams(spec.evaluation_params, multiturn),
  };
  if (spec.type === NodeType.TASK) {
    options.instructions = spec.instructions;
    options.outputLabel = spec.output_label;
  } else {
    options.criteria = spec.criteria;
  }
  if (multiturn && spec.turn_window != null) {
    options.turnWindow = [spec.turn_window[0], spec.turn_window[1]];
  }
  return options;
}

function metricOptions(
  spec: Record<string, any>,
  multiturn: boolean,
): Record<string, unknown> {
  return {
    name: spec.name,
    criteria: spec.criteria,
    evaluationSteps: spec.evaluation_steps,
    evaluationParams: deserializeEvalParams(spec.evaluation_params, multiturn),
    threshold: spec.threshold,
    strictMode: spec.strict_mode,
  };
}

function buildGEval(spec: any, multiturn: boolean): ChildMetric {
  const options = metricOptions(spec, multiturn);
  if (!options.name) {
    throw new Error("A 'geval' verdict child requires a 'name'.");
  }
  if (
    !Array.isArray(options.evaluationParams) ||
    options.evaluationParams.length === 0
  ) {
    throw new Error(
      "A 'geval' verdict child requires a non-empty 'evaluation_params'.",
    );
  }
  const cls = multiturn ? ConversationalGEval : GEval;
  return new cls(options as any);
}

function buildMetric(spec: any, multiturn: boolean): ChildMetric {
  const metricClass = spec.metric_class;
  if (typeof metricClass !== "string" || metricClass.length === 0) {
    throw new Error("A 'metric' verdict child requires a 'metric_class'.");
  }
  const cls = METRIC_CLASS_REGISTRY[metricClass];
  if (cls == null) {
    throw new Error(
      `Unknown metric_class '${metricClass}'. Register it first with ` +
        "registerMetricClass() so it can be reconstructed.",
    );
  }
  return new cls(metricOptions(spec.kwargs ?? {}, multiturn));
}

interface BuildOptions {
  multiturn: boolean;
  /** How a `metric`-typed verdict child is rebuilt from its spec. */
  buildChildMetric: (spec: any, multiturn: boolean) => ChildMetric;
}

function buildNodes(
  nodesSpec: Record<string, any>,
  options: BuildOptions,
): AnyNode[] {
  if (nodesSpec == null || Object.keys(nodesSpec).length === 0) {
    throw new Error(
      "Invalid DAG document: 'nodes' must be a non-empty object.",
    );
  }
  const validTypes: string[] = Object.values(NodeType);
  for (const [id, spec] of Object.entries(nodesSpec)) {
    if (spec == null || spec.type == null) {
      throw new Error(`Node '${id}' is missing required 'type' field.`);
    }
    if (!validTypes.includes(spec.type)) {
      throw new Error(
        `Node '${id}' has unknown type '${spec.type}'. Expected one of: ${validTypes.join(", ")}.`,
      );
    }
  }

  const referenced = collectReferencedIds(nodesSpec);
  const rootIds = Object.keys(nodesSpec).filter((id) => !referenced.has(id));
  if (rootIds.length === 0) {
    throw new Error(
      "No root nodes detected (every node is referenced as a child); " +
        "graph would be empty or contain a cycle.",
    );
  }

  const classes = NODE_CLASSES[options.multiturn ? "multi" : "single"];
  const built = new Map<string, AnyNode>();

  const build = (id: string, stack: Set<string>): AnyNode => {
    const existing = built.get(id);
    if (existing) return existing;
    if (stack.has(id)) {
      throw new Error(`Cycle detected in DAG refs involving node '${id}'.`);
    }
    const spec = nodesSpec[id];
    if (spec == null) {
      throw new Error(`Reference to unknown node id '${id}'.`);
    }

    stack.add(id);
    const cls = classes[spec.type as NodeType];
    let node: AnyNode;
    if (spec.type === NodeType.VERDICT) {
      node = buildVerdict(spec, cls, stack, build, options);
    } else {
      const children = (spec.children ?? []).map((childId: string) =>
        build(childId, stack),
      );
      node = new cls(nodeOptions(spec, options.multiturn)) as ExecutableNode;
      (node as ExecutableNode).children.push(...children);
    }
    stack.delete(id);
    built.set(id, node);
    return node;
  };

  return rootIds.map((id) => build(id, new Set()));
}

function buildVerdict(
  spec: any,
  cls: any,
  stack: Set<string>,
  build: (id: string, stack: Set<string>) => AnyNode,
  options: BuildOptions,
): VerdictLikeNode {
  if (spec.score != null) {
    return new cls({ verdict: spec.verdict, score: spec.score });
  }
  const childSpec = spec.child;
  if (childSpec == null || childSpec.type == null) {
    throw new Error(
      "VerdictNode spec must have either 'score' or a 'child' object.",
    );
  }
  let child: AnyNode | ChildMetric;
  switch (childSpec.type) {
    case ChildType.NODE:
      if (typeof childSpec.ref !== "string") {
        throw new Error("VerdictNode child of type 'node' requires 'ref'.");
      }
      child = build(childSpec.ref, stack);
      break;
    case ChildType.GEVAL:
      child = buildGEval(childSpec, options.multiturn);
      break;
    case ChildType.METRIC:
      child = options.buildChildMetric(childSpec, options.multiturn);
      break;
    default:
      throw new Error(
        `VerdictNode child has unknown type '${childSpec.type}'. ` +
          `Expected one of: ${Object.values(ChildType).join(", ")}.`,
      );
  }
  return new cls({ verdict: spec.verdict, child });
}

export function dagFromDict(
  data: Record<string, any>,
  multiturn = false,
): DeepAcyclicGraph {
  if (data == null || data.nodes == null) {
    throw new Error(
      "Invalid DAG document: expected an object with a 'nodes' key.",
    );
  }
  const rootNodes = buildNodes(data.nodes, {
    multiturn,
    buildChildMetric: buildMetric,
  });
  return new DeepAcyclicGraph({ rootNodes });
}

export function dagFromJson(json: string, multiturn = false): DeepAcyclicGraph {
  return dagFromDict(JSON.parse(json), multiturn);
}

// ----------------------------------------------------- Confident AI payloads

/**
 * The upload shape: identical to {@link dagToDict} except a metric child is
 * referenced by name, since the platform stores metrics separately.
 */
export function serializeDagToPayload(
  dag: DeepAcyclicGraph,
): Record<string, unknown> {
  return serializeDag(dag, (child, ids) =>
    isNode(child)
      ? { type: ChildType.NODE, ref: ids.get(child) }
      : { type: ChildType.METRIC, metric_name: child.name },
  );
}

export function constructDagUploadPayload(
  name: string,
  dag: DeepAcyclicGraph,
  multiTurn = false,
): Record<string, unknown> {
  return {
    name,
    algorithm: "DAG",
    multiTurn,
    dag: serializeDagToPayload(dag),
  };
}

/**
 * Rebuild a DAG from a Confident AI payload. Metric children are referenced by
 * name only; the TypeScript G-Eval metrics have no `pull()`, so those must be
 * re-attached in code.
 */
export function buildDagFromPayload(
  payload: Record<string, any>,
  multiturn = false,
): DeepAcyclicGraph {
  if (payload == null || payload.nodes == null) {
    throw new Error(
      "Invalid DAG document: expected an object with a 'nodes' key.",
    );
  }
  const rootNodes = buildNodes(payload.nodes, {
    multiturn,
    buildChildMetric: (spec) => {
      throw new Error(
        `Metric child '${spec.metric_name ?? spec.metric_class}' cannot be pulled: ` +
          "resolving a metric by name is not supported in the TypeScript SDK. " +
          "Rebuild that branch in code instead.",
      );
    },
  });
  return new DeepAcyclicGraph({ rootNodes });
}
